torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp

```diff
@@ -511,6 +552,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
+    new_value->copyMetadata(v);
+    return new_value;
+  } else {
+    return remap_values_in_observer_subgraph_[v];
+  }
+}
+
+void ProcessWeightObserverSubgraph::cloneNodeInGraph(
+    Node* node,
+    std::shared_ptr<Graph>& g) {
+  auto* block = g->block();
+  auto value_fn = [&](Value* v) { return updateInputValueInGraph(v, g); };
+
+  auto new_node = block->appendNode(g->createClone(node, value_fn));
+  for (size_t i = 0; i < node->outputs().size(); ++i) {
+    auto oo = node->outputs()[i];
+    auto no = new_node->outputs()[i];
+    remap_values_in_observer_subgraph_[oo] = no;
+  }
+}
+void ProcessWeightObserverSubgraph::buildObserverSubgraph(
+    const std::vector<Node*>& weight_subgraph,
+    std::shared_ptr<Graph> dest_graph) {
+  // Build weight subgraph
+  for (auto n : weight_subgraph) {
+    cloneNodeInGraph(n, dest_graph);
+  }
+  LintGraph(dest_graph);
+  // Add last node output value as subgraph output.
+  dest_graph->registerOutput(
+      updateInputValueInGraph(weight_subgraph.back()->output(0), dest_graph));
+  GRAPH_DUMP("New weight observer subgraph: ", dest_graph);
+}
+
+void ProcessWeightObserverSubgraph::findSubgraphNodes(
+    const std::vector<Value*>& inputs,
+    std::vector<Node*>& weight_subgraph) {
+  for (auto v : inputs) {
+    auto n = v->node();
+    if (!hitGraphInput(n->output(0))) {
+      weight_subgraph.push_back(n);
+      findSubgraphNodes(n->inputs().vec(), weight_subgraph);
+    }
+  }
+}
+
+void ProcessWeightObserverSubgraph::extractAndRunWeightObserver(
+    Module& module,
+    Value* self,
+    Value* weight_value,
+    std::unordered_map<Value*, std::vector<Node*>>& weight_to_subgraph_nodes) {
+  Node* observer = weight_value->node();
+  Graph* g = observer->owningGraph();
+  std::vector<Node*> weight_subgraph;
+
+  // If the graph was already visited, return list of relevant nodes directly.
+  // Multiple module instances can share the same graph code, so we don't need
+  // to re-run the extraction process.
+  if (weight_to_subgraph_nodes.count(weight_value)) {
+    weight_subgraph = weight_to_subgraph_nodes[weight_value];
+  } else {
+    // Extract the subgraph nodes.
+    weight_subgraph.push_back(observer);
+    // Track values in the subgraph for which producer needs to be found.
+    std::vector<Value*> curr_node_inputs;
+    for (auto v : observer->inputs()) {
+      curr_node_inputs.push_back(v);
+    }
+    findSubgraphNodes(observer->inputs().vec(), weight_subgraph);
+
+    // Reverse to traverse subgraph in correct direction
+    std::reverse(weight_subgraph.begin(), weight_subgraph.end());
+    weight_to_subgraph_nodes[weight_value] = weight_subgraph;
+  }
+
+  auto observer_subgraph = std::make_shared<Graph>();
+  auto build_observer_graph = [&](Function& func) {
+    buildObserverSubgraph(weight_subgraph, func.graph());
+  };
+  auto func = torch::make_unique<GraphFunction>(
+      "observer_subgraph", observer_subgraph, build_observer_graph);
```

Comment:
I think we can extract the building subgraph part to a separate class, and cache the graph function we build.

Reply:
I thought about it before but couldn't find a way to do it since this part depends on the build function for the subgraph. Functionality wise it shouldn't make a difference imo. Lmk if you have a different view.

Reply:
not sure I understand why this is not possible, I think if we can extract this as a separate class, or I feel we maybe able to extract this as a function, we can put this in jit, other people might be able to use it as well.